design/open/delta-diff.md

```diff
@@ -0,0 +1,168 @@
+2. Traverse through the [returned "commitInfo" entry vector ](https://github.com/delta-io/delta-rs/blob/main/rust/src/delta.rs#L888)
+starting from the **last** version of each history vector: 
+    1. While the returned entry versions **aren't** equal:
+        1. If the bigger version is the _"left"'s_ version, add the entry to the returned history list.
+        2. Traverse one version back of the bigger version entry's history.
+    2. While the versions > 0: 
+        1. Create a hash for the entry based on fields: `timestamp`, `operation`, `operationParameters` , and `operationMetrics` values.
+        2. Compare the hashes of the versions.
+        3. If they **aren't equal**, add the "left"'s entry to the returned history list, else break and **return the history vector**.
+        4. Traverse one version back in both vectors.
+3. Return an empty history vector.
+
+### Authentication
+
+The `delta-rs` package generates an S3 client which will be used to communicate back to lakeFS (through the S3 gateway).  
+In order for the S3 client to communicate with lakeFS (or S3 in general) it needs to pass an AWS Access Key Id and Secret access key.  
+Since applicative credentials are not obligatory, the users that sent the request for a diff might not have such, and even if they have,
+they cannot send them through the GUI (which is the UI we chose to implement this feature for the MVP).
+To overcome this scenario, we'll use special diff credentials as follows:
+1. User makes a Delta Diff request.
+2. The DiffService checks if the user has "diff credentials" in the DB:
+    1. If there are such credentials, it will use them.
+    2. If there aren't such, it will generate the credentials and save them: `{AKIA: DIFF-<>, SAK: <>}`. The `DIFF` prefix will be used to identify "diff credentials".
+3. The DiffService will pass the credentials to the Delta Diff plugin during the call as environment variables (or part of the `gRPC` call).
+
+### API
+
+- GET `/repositories/repo/{repo}/otf/refs/{left_ref}/diff/{right_ref}?type={diff_type}`
+    - Tagged as experimental
+    - **Response**:
+  ```json
+  [
+       {
+           "version": 1,
+           "timestamp":1515491537026,
+           "operation":"INSERT",
+           "operationContent":{
+               "operationParameters": {
+                  "mode":"Append",
+                  "partitionBy":"[]"
+                }
+      },
+      ...
+  ]
+  ```
+
+---
+
+### Packaging
+
+1. We will package the binary with the lakeFS release.
+2. We will also update the Dockerfile to include the binary within it.
+
+---
+
+## Metrics
+
+### Applicative Metrics
+- Diff runtime
+- The total number of versions returned - this is the number of files that were read from lakeFS. This is basically the size of a Delta Log. We can use it later on to optimize reading.
+
+### Business Statistics
+- The number of unique requests - get the number of Delta Lake users.
```

Comment:
Did you mean "get the number of Delta Lake **diff** users."? 
What defines a unique request?

Reply:
Both.
I would like to understand how many of our users use Delta Lake. I guess that not all of our Delta users will use the Delta Diff, but it might give us a better idea than the tools we currently have.
A unique request is defined by an installation id.

Reply:
Thanks, then consider using - 
```suggestion
- The number of unique requests by installation id - get the lower bound of the number of Delta Lake users.
```